### PR TITLE
15 add more termination criteria

### LINKED ISSUE #15: Add more termination criteria

### PR DESCRIPTION
**Pull request by dleidreiter**
Closes #15

### CODE REVIEW

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 1]
```diff
@@ -109,3 +110,184 @@ def check_termination(
...
+    Calculates the median Hausdorff distance between the expectation values in two population evaluation results
+
+    :arg result_1: first evaluation result to calculate the distance on
+    :type result_1: BasePopulationEvaluationResult
+    :arg result_2: second evaluation result to calculate the distance on
+    :type result_2: BasePopulationEvaluationResult
+    :return: the median Hausdorff distance between the expectation values in two population evaluation results
+    """
+
+    def distance(from_expectations: list[float], to_expectations: list[float]) -> float:
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:154]: These lists are of population size, right? So not too large I guess.

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 2]
```diff
@@ -109,3 +110,184 @@ def check_termination(
...
+        before this criterion chooses to terminate. If set to 0, this terminates the first time the change falls
+        below the threshold value. If set to 2 for example, this terminates the first time the change falls below
+        the threshold three consecutive times. Must be at least 0.
+    :type allowed_consecutive_violations: int
+    """
+
+    def __init__(self, minimum_change: float, allowed_consecutive_violations: int):
+        super().__init__()
+        self._minimum_change: float = minimum_change
+        self._allowed_consecutive_violations: int = allowed_consecutive_violations
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:187]: Maybe verify that `allowed_consecutive_violations >= 0`?

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 3]
```diff
@@ -109,3 +110,184 @@ def check_termination(
...
+
+        if self._last_population_evaluation is not None:
+            hausdorff_distance: float = _median_hausdorff_distance_by_expectation_value(
+                self._last_population_evaluation, population_evaluation
+            )
+            best_individual_distance: float = abs(
+                self._last_population_evaluation.best_expectation_value - population_evaluation.best_expectation_value
+            )
+            distance: float = max(hausdorff_distance, best_individual_distance)
+            last_population_median_expectation: float = cast(
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:272]: It doesn't really matter performwise, but you could also keep this value in a field the same way as `_last_population_evaluation`. I would think is is a tiny bit more intuitive but I don't care much either way.

**Review by svenpruefer — APPROVED** [review]
Thank you, this looks good to me!

File: queasars/minimum_eigensolvers/base/evolving_ansatz_minimum_eigensolver.py [hunk 4]
```diff
@@ -253,14 +254,11 @@ def result_callback(evaluation_result: BasePopulationEvaluationResult) -> None:
...
-            self.logger.info("Expectation value of best individual found so far: %f" % current_best_expectation_value)
+            self.logger.info("Current best expectation value: %f" % evaluation_result.best_expectation_value)
             filtered_expectations = [
                 expectation for expectation in evaluation_result.expectation_values if expectation is not None
             ]
-            self.logger.info(
-                "Average expectation value in the population currently: %f"
-                % (sum(filtered_expectations) / len(filtered_expectations))
-            )
+            self.logger.info("Median expectation value in the population currently: %f" % median(filtered_expectations))
```
[the remark this answers is not in the document]
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/evolving_ansatz_minimum_eigensolver.py]: Ok, i adjusted the log message and added a logging call for the mean expectation value.

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 1 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py:154]: Exactly. 👍

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 2 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py:189]: Done.

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 3 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py:276]: Since it is not super important to you either way, i will just leave it as is.

**Pull request merged**